src/main/java/com/salesforce/dynamodbv2/mt/util/RecordsCache.java

```diff
@@ -0,0 +1,276 @@
+    private final LinkedList<ShardLocation> insertionOrder;
+    // looks for mutating lock
+    private final ReadWriteLock lock;
+    // size of cache >= 0
+    private long recordsByteSize;
+
+    RecordsCache(long maxRecordsByteSize) {
+        this.maxRecordsByteSize = maxRecordsByteSize;
+        this.segments = new HashMap<>();
+        this.insertionOrder = new LinkedList<>();
+        this.lock = new ReentrantReadWriteLock();
+        this.recordsByteSize = 0L;
+    }
+
+    /**
+     * Returns the cached segment that contains the given shard location. If no such segment exists, returns empty.
+     *
+     * @param shardLocation Shard location for which to return the cached segment that contains it.
+     * @return Segment that contains the given location or empty.
+     */
+    List<Record> getRecords(ShardLocation shardLocation, int limit) {
+        assert shardLocation != null && limit > 0;
+
+        final Lock readLock = lock.readLock();
+        readLock.lock();
+        try {
+            final ShardId shardId = shardLocation.getShardId();
+            final NavigableMap<SequenceNumber, Segment> shardCache = segments.get(shardId);
+            if (shardCache == null) {
+                // nothing cached for the requested shard
+                return Collections.emptyList();
+            }
+
+            final SequenceNumber sequenceNumber = shardLocation.getSequenceNumber();
+            final Entry<SequenceNumber, Segment> entry = shardCache.floorEntry(sequenceNumber);
+            if (entry == null) {
+                // no segment with requested or smaller sequence number
+                return Collections.emptyList();
+            }
+
+            final Segment segment = entry.getValue();
+            if (segment.getEnd().precedes(sequenceNumber)) {
+                // preceding segment does not include requested sequence number
+                return Collections.emptyList();
+            }
+
+            // preceding segment contains (some) records for the requested sequence number
+            final List<Record> records = new ArrayList<>(limit);
+            addAll(records, segment.getRecords(sequenceNumber), limit);
+
+            // keep going through adjacent segments (if present), until limit is reached
+            while (records.size() < limit) {
+                final Segment next = shardCache.get(segment.getEnd().next());
+                if (next == null) {
+                    break;
+                }
+                addAll(records, next.getRecords(), limit);
+            }
+
+            return records;
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    // Should we bring back segment merging to avoid cache fragmentation?
+    void putRecords(ShardLocation shardLocation, List<Record> records) {
+        assert shardLocation != null && records != null && !records.isEmpty();
+
+        final Lock writeLock = lock.writeLock();
```

Comment:
is there any potential of web requests waiting on this lock? If so, it should be timed, and returned as metadata back to the client (ZOS) in whatever Result object is being passed back. When we cross over SLA's, we need to be able to track down the bottlenecks, and I'm worried newrelic isn't granular enough to do that for us, with its transaction sampling

Reply:
Great suggestion. The same holds for the shard-level lock (assuming we keep that).